### ISSUE TITLE
[rllib] test reward much lower than training reward on parametric DQN

### ISSUE THREAD

**Issue by laphang** [ray-project/ray#15162]
### What is the problem?

I slightly modified the parametric cartpole example to load back and evaluate the trained model, but for DQNs I get much smaller test rewards compared to training rewards. (For PPO, I get test rewards of 200 as expected). 

```
== Status ==
Memory usage on this node: 2.5/15.1 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/8 CPUs, 0/0 GPUs, 0.0/8.15 GiB heap, 0.0/2.78 GiB objects
Result logdir: /home/ec2-user/ray_results/DQN
Number of trials: 1/1 (1 TERMINATED)
+-----------------------------+------------+-------+--------+------------------+-------+----------+----------------------+----------------------+--------------------+
| Trial name                  | status     | loc   |   iter |   total time (s) |    ts |   reward |   episode_reward_max |   episode_reward_min |   episode_len_mean |
|-----------------------------+------------+-------+--------+------------------+-------+----------+----------------------+----------------------+--------------------|
| DQN_pa_cartpole_3bfed_00000 | TERMINATED |       |     43 |          171.998 | 43000 |   199.75 |                  200 |                  177 |             199.75 |
+-----------------------------+------------+-------+--------+------------------+-------+----------+----------------------+----------------------+--------------------+

Evaluation rewards:
12.0
12.0
19.0
35.0
13.0
10.0
13.0
22.0
22.0
25.0
```

*Ray version and other system information (Python version, TensorFlow version, OS):* Ray 1.1.0, Python 3.6.13, TF 2.1.3, linux

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

```

import argparse
import os

import ray
from ray import tune
from ray.rllib.examples.env.parametric_actions_cartpole import \
    ParametricActionsCartPole
from ray.rllib.examples.models.parametric_actions_model import \
    ParametricActionsModel, TorchParametricActionsModel
from ray.rllib.models import ModelCatalog
from ray.rllib.utils.test_utils import check_learning_achieved
from ray.tune.registry import register_env
from ray.rllib.agents import dqn, ppo
from ray.rllib.agents.registry import get_agent_class

parser = argparse.ArgumentParser()
parser.add_argument("--run", type=str, default="PPO")
parser.add_argument("--torch", action="store_true")
parser.add_argument("--as-test", action="store_true")
parser.add_argument("--stop-iters", type=int, default=100)
parser.add_argument("--stop-reward", type=float, default=199.0)
parser.add_argument("--stop-timesteps", type=int, default=100000)

if __name__ == "__main__":
    args = parser.parse_args()
    ray.init()

    register_env("pa_cartpole", lambda _: ParametricActionsCartPole(2))
    ModelCatalog.register_custom_model(
        "pa_model", TorchParametricActionsModel
        if args.torch else ParametricActionsModel)

    
    if args.run =='DQN':
        config = dqn.DEFAULT_CONFIG.copy()
        config['env'] = 'pa_cartpole'
        config['model']["custom_model"]="pa_model" 
        config['num_workers']=0
        config['hiddens'] = []
        config['model']['fcnet_hiddens'] = [256]
    elif args.run =='PPO':
        config = ppo.DEFAULT_CONFIG.copy()
        config['env'] = 'pa_cartpole'
        config['model']["custom_model"]="pa_model" 
        config['num_workers']=7
    else:
        raise ValueError()    

    stop = {
        "training_iteration": args.stop_iters,
        "timesteps_total": args.stop_timesteps,
        "episode_reward_mean": args.stop_reward,
    }

    results = tune.run(args.run, stop=stop, config=config, checkpoint_freq=20, checkpoint_at_end=True, verbose=2)

    if args.as_test:
        check_learning_achieved(results, args.stop_reward)
        
    #load checkpoint
    m = "episode_reward_mean"
    checkpoint_path = results.get_trial_checkpoints_paths(results.get_best_trial(m, mode="max"), metric=m)[-1][0]
    print(checkpoint_path)
    cls = get_agent_class(args.run)
    config['num_workers']=0
    agent = cls(env='pa_cartpole', config=config)
    agent.restore(checkpoint_path)
    
    #do evaluation 
    print("Evaluation rewards:")
    env = ParametricActionsCartPole(2)
    for i in range(10):        
        episode_reward = 0
        done = False
        obs = env.reset()
        while not done:
            action = agent.compute_action(obs)
            obs, reward, done, info = env.step(action)
            episode_reward += reward
        print(episode_reward)
    

    ray.shutdown()

```

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by piscesvivian** [comment]
I have the same problem on CartPole APEX

**Comment by laphang** [comment]
Just adding an update to this, after noticing this issue https://github.com/ray-project/ray/issues/16065 on the DQN for the simple cartpole, and the possible fix for that issue here https://github.com/ray-project/ray/pull/16354, and whether it would have any impact on the parametric cartpole. 

Firstly, I was able to to reproduce the behaviour in my original post (which used ray 1.1.0) on the latest ray release 1.4.0 (ie for the parametric cartpole, PPO test reward around 200, and DQN test reward was very low). 

However, I find that if I use the latest daily builds (https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-2.0.0.dev0-cp36-cp36m-manylinux2014_x86_64.whl) in order to include the fix #16354, then I find that for the parametric cartpole, both PPO and DQN has very low test rewards, so it seems to have not fixed DQN and has broken PPO. 

For a DQN on the simple cartpole, I did find that the test reward was very low in 1.4.0, but was fixed (ie. around 200) when using the latest daily build, so #16354 seemed to work for the simple cartpole. 

@sven1977 I was wondering if you had any thoughts on DQN on parametric cartpole?
